In [116]:
import ROOT
import os

In [117]:
# !combine --help

### Important:
signal strength r = (number of signal events / number of expected signal events).    
It scales the signal up and down in such a way after considering all the nuisance parameters. Signal strength * signal = eg. expectation of number of events in the cut and count physics process.
data_obs = e.g., observed number of events in the data
POI = e.g., ttgamma crossection which is burried in signal histogram. 

### FitDiagnostics:
performs maximum likelihood fits to extract the signal yield and provide diagnostic tools such as pre and post-fit models and correlations

### MultiDimFit:
perform maximum likelihood fits in multiple parameters and likelihood scans

### Doing exercise for misIDEle

In [118]:
# myfile = ROOT.TFile("MyFiles/pho_looseCRe2e0plots_ele_2016/myMisIDEle_Prefit.root")
# myfile.ls()
# MyDataHist = myfile.Get("myData")
# MisIDEle   = myfile.Get("myMisIDEle")
# MisIDEle.SetLineWidth(2)
# MisIDEle.SetLineColor(2)
# OtherHist  = myfile.Get("myBackground")
# OtherHist.SetLineWidth(2)
# OtherHist.SetLineColor(3)

# c = ROOT.TCanvas()
# MyDataHist.Draw("hist")
# MisIDEle.Draw("same") # red
# OtherHist.Draw("same") # yellow
# c.Draw()


In [119]:
# !text2workspace.py datacard_AR.txt

In [120]:
# f_shapes_TH1 = ROOT.TFile("datacard_AR.root","READ")
# w_shapes_TH1 = f_shapes_TH1.Get("w")
# w_shapes_TH1.Print()

In [121]:
# !python ../test/printWorkspaceNormalisations.py datacard_AR.root

In [122]:
# !combine -M MultiDimFit -n AR  datacard_AR.txt --redefineSignalPOIs r,mybackground_norm

In [123]:
!combine -M FitDiagnostics -n AR  datacard_AR.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          757.540155      27.523447       total sum                     
0          758.000000      27.531800       Unweighted events, alpha=0.999393
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          803.027725      28.337744       total sum                     
1          803.000000      28.337255       Unweighted events, alpha=1.000035
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          925.131241      30.415970       total sum                     
2      

In [124]:
!combine -M FitDiagnostics -n CR123 datacard_CR123.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          617.905571      24.857706       total sum                     
0          618.000000      24.859606       Unweighted events, alpha=0.999847
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          624.114464      24.982283       total sum                     
1          624.000000      24.979992       Unweighted events, alpha=1.000183
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          731.822906      27.052226       total sum                     
2      

In [125]:
!combine -M FitDiagnostics -n CR1 datacard_CR1.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          475.777184      21.812317       total sum                     
0          476.000000      21.817424       Unweighted events, alpha=0.999532
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          475.096313      21.796704       total sum                     
1          475.000000      21.794495       Unweighted events, alpha=1.000203
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          600.655792      24.508280       total sum                     
2      

In [126]:
!combine -M FitDiagnostics -n CR2 datacard_CR2.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          107.881416      10.620062       total sum                     
0          103.000000      10.148892       Unweighted events, alpha=1.047392
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          95.907937       9.793260        total sum                     
1          96.000000       9.797959        Unweighted events, alpha=0.999041
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          93.435108       9.666184        total sum                     
2      

In [127]:
!combine -M FitDiagnostics -n CR3 datacard_CR3.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          84.021584       9.524956        total sum                     
0          78.000000       8.831761        Unweighted events, alpha=1.077200
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          90.842171       9.531116        total sum                     
1          91.000000       9.539392        Unweighted events, alpha=0.998266
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          89.861055       9.479507        total sum                     
2      

In [128]:
!combine -M FitDiagnostics -n CR4 datacard_CR4.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          130.149635      11.408314       total sum                     
0          130.000000      11.401754       Unweighted events, alpha=1.001151
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          145.243824      12.051715       total sum                     
1          145.000000      12.041595       Unweighted events, alpha=1.001682
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          251.466347      15.857690       total sum                     
2      

In [129]:
!combine -M FitDiagnostics -n CR5 datacard_CR5.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          64.665924       8.041513        total sum                     
0          65.000000       8.062258        Unweighted events, alpha=0.994860
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          70.980239       8.424977        total sum                     
1          71.000000       8.426150        Unweighted events, alpha=0.999722
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          77.983253       8.830813        total sum                     
2      

In [130]:
!combine -M FitDiagnostics -n CR6 datacard_CR6.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          23.988344       4.897790        total sum                     
0          24.000000       4.898979        Unweighted events, alpha=0.999514
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          28.510343       5.339508        total sum                     
1          29.000000       5.385165        Unweighted events, alpha=0.983115
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          17.051010       4.129287        total sum                     
2      

In [131]:
!combine -M FitDiagnostics -n CR7 datacard_CR7.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          16.564990       4.070011        total sum                     
0          17.000000       4.123106        Unweighted events, alpha=0.974411
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          33.239134       5.765339        total sum                     
1          33.000000       5.744563        Unweighted events, alpha=1.007246
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          22.462335       4.739445        total sum                     
2      

In [132]:
!combine -M FitDiagnostics -n SR8 datacard_SR8.txt 

 <<< Combine >>> 
>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Channel mybin will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmybin
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          68.316554       8.265383        total sum                     
0          68.000000       8.246211        Unweighted events, alpha=1.004655
  => Total parameter prop_binmybin_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          94.248138       9.708148        total sum                     
1          94.000000       9.695360        Unweighted events, alpha=1.002640
  => Total parameter prop_binmybin_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          79.073237       8.892313        total sum                     
2      

In [133]:
# f_input_shapes_TH1 = ROOT.TFile("simple-shapes-TH1_input.root","READ")
# f_input_shapes_TH1.ls()


In [134]:
# h_background = f_input_shapes_TH1.Get("background")
# h_background.SetLineColor(ROOT.kBlack)
# print h_background.Integral(-1,-1)

# h_background_alphaUp = f_input_shapes_TH1.Get("background_alphaUp")
# h_background_alphaUp.SetLineColor(ROOT.kBlue)
# print h_background_alphaUp.Integral(-1,-1)

# h_background_alphaDown = f_input_shapes_TH1.Get("background_alphaDown")
# h_background_alphaDown.SetLineColor(ROOT.kRed)
# print h_background_alphaDown.Integral(-1,-1)

# c = ROOT.TCanvas()
# h_background_alphaDown.Draw()
# h_background.Draw("same")
# h_background_alphaUp.Draw("same")

# c.Draw()

In [135]:
# h_signal = f_input_shapes_TH1.Get("signal")
# h_signal.SetLineColor(ROOT.kBlack)
# print h_signal.Integral()

# h_signal_sigmaUp = f_input_shapes_TH1.Get("signal_sigmaUp")
# h_signal_sigmaUp.SetLineColor(ROOT.kBlue)
# print h_signal_sigmaUp.Integral()

# h_signal_sigmaDown = f_input_shapes_TH1.Get("signal_sigmaDown")
# h_signal_sigmaDown.SetLineColor(ROOT.kRed)
# print h_signal_sigmaDown.Integral()

# c = ROOT.TCanvas()
# h_signal_sigmaDown.Draw()
# h_signal.Draw("same")
# h_signal_sigmaUp.Draw("same")

# c.Draw()

The combine tool converts datacards into a RooWorkspace to build the likelihood

In [136]:
# !text2workspace.py ttgamma_sys.txt

In [137]:
# f_shapes_TH1 = ROOT.TFile("ttgamma_sys.root","READ")
# w_shapes_TH1 = f_shapes_TH1.Get("w")
# w_shapes_TH1.Print()

The workspace contains "model_s" i.e. the likelihood function, which is composed of the observable "CMS_th1x" which is split in several channels, the POI "r" and several nuisance parameters. Lets see how combine smoothly morphs the shape of the background as a function of the nuisance parameter alpha:


In [138]:
# shapeBkg = w_shapes_TH1.pdf("shapeBkg_mybin_background_morph")
# th1x = w_shapes_TH1.var("CMS_th1x")
# plot_th1x = th1x.frame()
# alpha = w_shapes_TH1.var("alpha")
# alpha.setVal(0.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kBlack))
# alpha.setVal(0.5)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kGreen))
# alpha.setVal(1.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kRed))
# alpha.setVal(2.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kOrange))

# c2 = ROOT.TCanvas()
# plot_th1x.Draw()
# c2.Draw()

In [139]:
# shapeSig = w_shapes_TH1.pdf("shapeSig_mybin_signal_morph")
# th1x = w_shapes_TH1.var("CMS_th1x")
# plot_th1x = th1x.frame()
# sigma = w_shapes_TH1.var("sigma")
# sigma.setVal(0.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kBlack))
# sigma.setVal(0.5)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kGreen))
# sigma.setVal(1.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kRed))
# sigma.setVal(2.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kOrange))

# c2 = ROOT.TCanvas()
# plot_th1x.Draw()
# c2.Draw()

Prints the workspace normalization 

In [140]:
# !python ../test/printWorkspaceNormalisations.py ttgamma_sys.root

If we redefine POI then it will be float and can take any values. If we have things in datacard then everything will be used to produce the best r value.  
The option "--run both" tells combine to run both expected and observed limits.You can see we have computed an observed limit on r, and also the nominal expected, and +/- 1 and 2 sigma expected limits. [-s -1 == random seed]

In [141]:
# !combine -M MultiDimFit -n Nabin ttgamma_sys.txt

In [142]:
# !combine -n LimitTestNabin -M AsymptoticLimits ttgamma_sys.root --run both

In [143]:
# asymptoticfile = ROOT.TFile("higgsCombineLimitTestNabin.AsymptoticLimits.mH120.root")
# limit = asymptoticfile.Get("limit")
# for i in xrange(limit.GetEntries()):
#     limit.GetEntry(i)
#     print i,limit.mh,limit.limit

1-sigma ==  68%, 2-sigma == 95%, 3-sigma == 99.7%. The higher sigma value means that the discovery is less likely to be accidently a mistake or 'random chance'.   
How do we test hypothesis?  
Null hypothesis  means A and B are not connected. Research hypothesis means A and B are connected. First assume Null hypothesis is true. Then calculate the probability of obtaining data as extreme as the kind they are observing. 
P-value close to zero means Null hypothesis is false, however, larger p-value means Null hypothesis is true. For example, p-value = 5% means that chance that the data sample being tested is drawn from where Null hypothesis is true.  
Next example, the chances of a result being greater than 1-sigma is 15.9%, 2-sigma is 2.3%, 3-sigma is 0.2%. 